python/ray/dag/tests/experimental/test_execution_schedule.py

```diff
@@ -0,0 +1,1017 @@
+        branch2 = b.no_op.bind(inp)
+        branch2.with_type_hint(TorchTensorType(transport="nccl"))
+        branch3 = c.no_op.bind(inp)
+        branch3.with_type_hint(TorchTensorType(transport="nccl"))
+        dag = MultiOutputNode(
+            [
+                a.no_op.bind(branch3),
+                b.no_op.bind(branch1),
+                c.no_op.bind(branch2),
+            ]
+        )
+
+    compiled_dag = dag.experimental_compile()
+
+    a_expected_schedule = [
+        (0, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.WRITE),
+        (1, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.WRITE),
+    ]
+    b_expected_schedule = [
+        (0, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.WRITE),
+        (1, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.WRITE),
+    ]
+    c_expected_schedule = [
+        (0, DAGNodeOperationType.READ),
+        (0, DAGNodeOperationType.COMPUTE),
+        (0, DAGNodeOperationType.WRITE),
+        (1, DAGNodeOperationType.READ),
+        (1, DAGNodeOperationType.COMPUTE),
+        (1, DAGNodeOperationType.WRITE),
+    ]
+
+    a_schedule = compiled_dag.actor_to_execution_schedule[a]
+    b_schedule = compiled_dag.actor_to_execution_schedule[b]
+    c_schedule = compiled_dag.actor_to_execution_schedule[c]
+
+    for schedule, expected_schedule in zip(
+        [a_schedule, b_schedule, c_schedule],
+        [a_expected_schedule, b_expected_schedule, c_expected_schedule],
+    ):
+        assert len(schedule) == len(expected_schedule)
+        for i, operation in enumerate(schedule):
+            assert operation.idx == expected_schedule[i][0]
+            assert operation.type == expected_schedule[i][1]
+
+    tensor_cpu = torch.zeros(10, 10)
+    ref = compiled_dag.execute(tensor_cpu)
+    tensors = ray.get(ref)
+    tensor_cuda = tensor_cpu.to("cuda:0")
+
+    assert len(tensors) == 3
+    for t in tensors:
+        assert torch.equal(t, tensor_cuda)
+
+    compiled_dag.teardown()
+
+
+def generate_dag_graph_nodes(local_idx, global_idx, actor_handle, requires_nccl):
+    graph_nodes = {}
+    for op_type in DAGNodeOperationType:
+        graph_nodes[op_type] = DAGOperationGraphNode(
+            DAGNodeOperation(local_idx, op_type),
+            global_idx,
+            actor_handle,
+            requires_nccl,
+        )
+    return graph_nodes
+
+
+class TestSelectNextNodes:
+    """
+    Test whether `_select_next_nodes` function selects the next nodes for
+    topological sort to generate execution schedule correctly.
+
+    global_idx: Each DAG node has a unique global index.
+    local_idx: The DAG node's index in the actor's `executable_tasks` list.
+    """
+
+    def test_two_candidates_on_same_actor(self):
```

Comment:
can you write a test to run the algorithm N times (with an example that has lots of ties) and make sure the order is always deterministic?

Reply:
Added https://github.com/ray-project/ray/pull/46911/commits/fc123cbb610f198a117f722d65013ec752e0047c